auth/api/iam/client.go

```diff
@@ -0,0 +1,88 @@
+/*
+ * Copyright (C) 2023 Nuts community
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <https://www.gnu.org/licenses/>.
+ *
+ */
+
+package iam
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"io"
+	"net/http"
+	"net/url"
+	"strings"
+
+	"github.com/nuts-foundation/go-did/did"
+	"github.com/nuts-foundation/nuts-node/core"
+)
+
+// HTTPClient holds the server address and other basic settings for the http client
+type HTTPClient struct {
+	httpClient core.HTTPRequestDoer
+}
+
+// NewHTTPClient creates a new api client.
+func NewHTTPClient(config core.ClientConfig) HTTPClient {
+	return HTTPClient{
+		httpClient: core.MustCreateHTTPClient(config, nil),
+	}
+}
+
+func (hb HTTPClient) clientWithBase(baseURL string) ClientInterface {
+	// can only be used for public APIs
+	response, err := NewClientWithResponses(baseURL, WithHTTPClient(hb.httpClient))
+	if err != nil {
+		panic(err)
+	}
+	return response
+}
+
+func (hb HTTPClient) OAuthAuthorizationServerMetadata(ctx context.Context, serverDID did.DID) (*OAuthAuthorizationServerMetadata, error) {
+	if serverDID.Method != "web" {
+		return nil, fmt.Errorf("unsupported DID method: %s", serverDID.Method)
+	}
+	// TODO: ignoring root web did for now. We can't use the generated client for that type of web:did :(
+	// TODO: use IssuerIdToWellKnown func when merged
+	serverURL := strings.ReplaceAll(serverDID.ID, ":", "/")
+	// remove % encodings
+	serverURL, err := url.QueryUnescape(serverURL)
+	if err != nil {
+		return nil, err
+	}
+
+	response, err := hb.clientWithBase(fmt.Sprintf("https://%s", serverURL)).OAuthAuthorizationServerMetadata(ctx, serverDID.String())
```

Comment:
serverURL should be the OAuth issuer identifier (before adding well-known).

from [RFC8414](https://www.rfc-editor.org/rfc/rfc8414.html#section-2):
> issuer
      REQUIRED.  The authorization server's issuer identifier, which is
      a URL that uses the "https" scheme and has no query or fragment
      components.  Authorization server metadata is published at a
      location that is ".well-known" according to [RFC 5785](https://www.rfc-editor.org/rfc/rfc5785) [[RFC5785](https://www.rfc-editor.org/rfc/rfc5785)]
      derived from this issuer identifier, as described in [Section 3](https://www.rfc-editor.org/rfc/rfc8414.html#section-3).
      The issuer identifier is used to prevent authorization server mix-
      up attacks, as described in "OAuth 2.0 Mix-Up Mitigation"
      [[MIX-UP](https://www.rfc-editor.org/rfc/rfc8414.html#ref-MIX-UP)].

`https` is already checked in `IssuerIdToWellKnown`. 
Query/frament are not allowed. Add to `IssuerIdToWellKnown`?

Reply:
Replaced with DIDToURL